compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUTileRootAndFuseProducerConsumer.cpp

```diff
@@ -0,0 +1,207 @@
+    rewriter.replaceOp(fusedResult->origConsumerOperand->getOwner(),
+                       fusedResult->tiledOps.front());
+
+    // The result of the fused conumers might themselved be slices of
+    // values produced by operations that implement the `TilingInterface`.
+    // Add these operations to the worklist.
+    addCandidateSlices(fusedResult->tiledAndFusedConsumerOperand->getOwner(),
+                       candidates);
+  }
+  return success();
+}
+
+static LogicalResult tileRootAndFuseProducerConsumer(IRRewriter &rewriter,
+                                                     TilingInterface rootOp,
+                                                     int64_t tilingLevel) {
+
+  SmallVector<OpFoldResult> tileSizes =
+      getLoweringConfig(rootOp).getTilingLevelSizes(rewriter, tilingLevel,
+                                                    rootOp);
+  int64_t numLoops = rootOp.getLoopIteratorTypes().size();
+  if (tileSizes.size() > numLoops)
+    return failure();
+
+  scf::SCFTilingOptions tilingOptions;
+  tilingOptions.setTileSizes(tileSizes);
+
+  scf::SCFTileAndFuseOptions tileAndFuseOptions;
+  tileAndFuseOptions.setTilingOptions(tilingOptions);
+
+  return tileRootAndFuseProducerConsumerUsingSCF(rewriter, rootOp,
+                                                 tileAndFuseOptions);
+}
+
+/// This pass starts with the first TilingInterface operation that has
+/// lowering_config attribute, tiles the op and fuses its  consumers and
+/// producers recursively. The `tilingLevel` must be specified. It picks the
+/// `tilingLevel`-th list as tiling sizes from lowering_config.
+struct LLVMCPUTileRootAndFuseProducerConsumer
+    : LLVMCPUTileRootAndFuseProducerConsumerBase<
+          LLVMCPUTileRootAndFuseProducerConsumer> {
+  LLVMCPUTileRootAndFuseProducerConsumer(int64_t tilingLevel = -1) {
+    this->tilingLevel.setValue(tilingLevel);
+  }
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<arith::ArithDialect, affine::AffineDialect,
+                    linalg::LinalgDialect, scf::SCFDialect,
+                    vector::VectorDialect>();
+  }
+
+  void runOnOperation() override;
+};
+
+void LLVMCPUTileRootAndFuseProducerConsumer::runOnOperation() {
+  MLIRContext *context = &getContext();
+  auto funcOp = getOperation();
+
+  IRRewriter rewriter(funcOp);
+  TilingInterface rootOp = nullptr;
+
+  // Find the first operation with the lowering_config attribute.
+  funcOp->walk([&](TilingInterface target) {
+    if (IREE::Codegen::LoweringConfigAttrInterface loweringConfig =
+            getLoweringConfig(target)) {
+      if (loweringConfig.hasTilingLevel(tilingLevel)) {
+        rootOp = target;
+        return WalkResult::interrupt();
+      }
+    }
+    return WalkResult::advance();
+  });
+
+  if (!rootOp) {
+    funcOp.emitError()
+        << "no operation found with the lowering_config attribute.\n";
+    return signalPassFailure();
+  }
```

Comment:
IMO, it is better to signal a failure when there are multiple ops that have lowering_config. It makes the definition of "rootOp` more clear.

Reply:
Or we can make `rootOp` the very first op in the graph (in order) with the lowering config attached. Do you sense an issue here?

Reply:
After thinking a little more, I think we should use the `getRootOperation` method because it is how we define root op in `KernelDispatch.cpp`. It makes the rootOp definition in CPU pipeline consistent. What do you think?

```cpp
SmallVector<Operation *> computeOps = getComputeOps(funcOp);
FailureOr<Operation *> maybeRootOp = getRootOperation(computeOps);
// check if the op has lowering_config or not.
```

Reply:
After looking into the functions, I found that the above function has nothing to do with `LoweringConfigAttributeInterface`(If we attach lowering config to pack/unpack, this might end up failing, though I think that shouldn't be the case).  I can create a modified `getComputeOps` that also checks for lowering config and pass it to the `getRootOperation` fn.  https://github.com/iree-org/iree/blob/3fd336f2850e3965a26fab27a70e160d87588552/compiler/src/iree/compiler/Codegen/Common/GPU/GPUApplyTilingLevel.cpp#L176

Reply:
I don't follow the issue... Why do we care about lowering config in `getComputeOps`?

The idea of the new pipeline flow is "only the root op has lowering config". The logic in my mind is:

1. We get the root op.
2. The root op should have lowering_config. If it does have a lowering_config, it is invalid. It violates the whole design idea.
3. Start everything from root op.

The way to get the root op is following what we've done in `KernelDispatch.cpp`. We don't care whether it has lowering_config or not when we're finding the root op.

The `getRootOperation` method defines what the root op is.

https://github.com/iree-org/iree/blob/e40e64e23e3eaca95b9474e15000349b2c5203fb/compiler/src/iree/compiler/Codegen/Utils/CPUUtils.h#L14-L20

It is true that the op with lowering config is the root op in the pipeline. But it is not the way that we define root op. It sounds fragile to me if we take a step forward that assumes the op with lowering config is the root op. The assumption is correct, but the logic builds on top of how we set lowering_config. I think my question is more like -- why can't we follow the same logic to get the root op?

Reply:
I see, makes a lot more sense now.